### PR TITLE
fix: OPTIC-683: Detect and convert large integers automatically in API responses before parsing as JSON to retain precision

### PR DESCRIPTION
**Pull request by bmartel**
### PR fulfills these requirements
- [X] Commit message(s) and PR title follows the format `[fix|feat|ci|chore|doc]: TICKET-ID: Short description of change made` ex. `fix: DEV-XXXX: Removed inconsistent code usage causing intermittent errors`
- [X] Tests for the changes have been added/updated (for bug fixes/features)
- [X] Docs have been added/updated (for bug fixes/features)
- [X] Best efforts were made to ensure docs/code are concise and coherent (checked for spelling/grammatical errors, commented out code, debug logs etc.)
- [X] Self-reviewed and ran all changes on a local instance (for bug fixes/features)

#### Change has impacts in these area(s)
_(check all that apply)_
- [ ] Product design
- [ ] Backend (Database)
- [ ] Backend (API)
- [X] Frontend

### Describe the reason for change
When data within JSONB fields of the database would store large integers from other processes, when it came time to displaying these visually it would be done with loss of precision. The JS support for large integers is not very good being it only actually stores floats and has only 32bit support. Therefore to fix the display issues of these, I found that we were already performing a 2 stage response with all api calls from datamanager, therefore I added a method to replace large integers that are present in the textual data as the reviver function to JSON.parse running. This solves the issue and ultimately ensures we do not modify unknowingly data stored on the server while still retaining the visual accuracy for consumption in the UI.

_Before_
image

_After_
image

#### What alternative approaches were there?
- Support BigInt in JS.
    - Problem here is it doesn't actually solve the original problem.
    - Can be explicitly supported in addition to this solution but this requires that all other Number values be converted to BigInt to be able to perform mathematical operations on them.
    - Many caveats to cover which ultimately the solution we need is to be as low risk to data corruption as possible.

#### What feature flags were used to cover this change?
Don't have one yet, will entertain one if we feel it necessary.

### Does this PR introduce a breaking change?
_(check only one)_
- [ ] Yes, and covered entirely by feature flag(s)
- [ ] Yes, and covered partially by feature flag(s)
- [X] No
- [ ] Not sure (briefly explain the situation below)

### What level of testing was included in the change?
_(check all that apply)_
- [ ] e2e
- [ ] integration
- [X] unit

### Which logical domain(s) does this change affect?
DataManager API proxy.

### CODE REVIEW

**Review by yyassi-heartex — APPROVED** [review]

File: web/libs/datamanager/src/utils/__tests__/helpers.test.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
+import * as helpers from "../helpers";
+
+describe("helpers", () => {
+  const originalMaxSafeInteger = Number.MAX_SAFE_INTEGER;
+
+  beforeEach(() => {
+    // Mock Number.MAX_SAFE_INTEGER to a smaller value for testing purposes
```
Comment (bmartel) [inline web/libs/datamanager/src/utils/__tests__/helpers.test.ts]:
```suggestion
    // Mock Number.MAX_SAFE_INTEGER to a smaller value for testing purposes as NodeJS supports larger integers than Web.
```

**Comment by yyassi-heartex** [comment]
/selenium test
> '/selenium' is an unknown pull-request command.
> See '/help'

**Review by AndrejOros — APPROVED** [review]

File: web/libs/datamanager/src/utils/helpers.ts [hunk 2]
```diff
@@ -12,6 +12,35 @@ export const formDataToJPO = (formData: FormData) => {
   return formData;
 };
 
+/**
+ * Hydrate JSON values that are large integers to strings.
+ */
+export const jsonReviverWithBigInt = (_key: any, value: any, context: any) => {
```
Comment (hlomzik) [inline web/libs/datamanager/src/utils/helpers.ts]: easy fix for types problem:
```suggestion
export const jsonReviverWithBigInt = (_key: any, value: any, context?: any) => {
```

not all the browsers support context and while your code will work everywhere the typing for `parse()` has no `context` yet, so marking it as an optional will calm TS down.

File: web/libs/datamanager/src/utils/helpers.ts [hunk 3]
```diff
@@ -12,6 +12,35 @@ export const formDataToJPO = (formData: FormData) => {
...
+ * This is a workaround for the fact that JSON.parse does not directly support big integers and will
+ * immediately convert them to numbers (losing precision).
+ *
+ * ex. { "id": 12345678901234567890 } => { "id": "12345678901234567890" }
+ *     { "id": -12345678901234567890 } => { "id": "-12345678901234567890" }
+ *     { "meta": { "id": 12345678901234567890 } } => { "meta": { "id": "12345678901234567890" } }
+ *     { "meta": { "id": -12345678901234567890 } } => { "meta": { "id": "-12345678901234567890" } }
+ **/
+export const parseJson = <T = any>(jsonString: string): T => {
+  // @ts-ignore The types were complaining about the signature which I believe is due to the NodeJS types of JSON.parse getting in the way
```
Comment (hlomzik) [inline web/libs/datamanager/src/utils/helpers.ts]: and then we don't need this
```suggestion
```

**Review by hlomzik — APPROVED** [review]

**Pull request merged**